k8s.go

```diff
@@ -0,0 +1,643 @@
+	// compat adaptor to write the values in etcdv2 format.
+	client := compat.NewAdaptor(&etcdVarClient{vars: vars})
+	for _, kvp := range kvps {
+		if _, err := client.Apply(kvp); err != nil {
+			log.Error("Failed to convert k8s data to etcdv2 equivalent: %s = %s", kvp.Key, kvp.Value)
+		}
+	}
+}
+
+type IntervalProcessor struct {
+	interval int
+	k8scli   *k8sClient
+	ipam     *ipamCacheK8s
+}
+
+func (p *IntervalProcessor) IntervalLoop() error {
+	if err := p.k8scli.updatePrefix(); err != nil {
+		return err
+	}
+	if err := p.k8scli.initialNeighborConfigSetting(); err != nil {
+		return err
+	}
+	ippool, err := p.ipam.getIPPools()
+	if err != nil {
+		return err
+	}
+	p.ipam.lastIPPool = ippool
+	for {
+		log.Debug("polling")
+		p.ipam.sync()
+		p.k8scli.checkBGPConfig()
+		select {
+		case <-time.After(time.Duration(p.interval) * time.Second):
+			continue
+		}
+	}
+}
+
+type k8sClient struct {
+	node              string
+	server            *Server
+	k8scli            *kubernetes.Clientset
+	nodeBgpPeerClient resources.K8sNodeResourceClient
+	nodeBgpCfgClient  resources.K8sNodeResourceClient
+	lastBgpconfig     map[string]string
+}
+
+// create new Kubernetes client
+func NewK8sClient(s *Server) (*k8sClient, error) {
+	loadingRules := clientcmd.ClientConfigLoadingRules{}
+	config, err := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+		&loadingRules, &clientcmd.ConfigOverrides{}).ClientConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	// Create the clientset
+	cs, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+	return &k8sClient{
+		node:              os.Getenv(NODENAME),
+		server:            s,
+		k8scli:            cs,
+		nodeBgpPeerClient: resources.NewNodeBGPPeerClient(cs),
+		nodeBgpCfgClient:  resources.NewNodeBGPConfigClient(cs),
+	}, nil
+}
+
+func (c *k8sClient) updatePrefix() error {
+	var paths []*svbgptable.Path
+	node, err := c.k8scli.Nodes().Get(c.node, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	cidr := node.Spec.PodCIDR
+	path, err := c.server.makePath(cidr, false)
+	log.Debugf("Set prefix: %#v", path)
+	paths = append(paths, path)
+	if err = c.server.updatePrefixSet(paths); err != nil {
+		return err
+	}
+	if _, err := c.server.bgpServer.AddPath("", paths); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *k8sClient) initialNeighborConfigSetting() error {
+	bgpconfig, err := c.getBGPConfig()
+	if err != nil {
+		return err
+	}
+	c.lastBgpconfig = bgpconfig
+	neighborConfigs, err := c.getNeighborConfigs(c.lastBgpconfig)
+	if err != nil {
+		return err
+	}
+	for _, n := range neighborConfigs {
+		if err = c.server.bgpServer.AddNeighbor(n); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// getNeighborConfigs returns the complete list of BGP neighbor configuration
+// which the node should peer.
+func (c *k8sClient) getNeighborConfigs(bgpconfig map[string]string) ([]*svbgpconfig.Neighbor, error) {
```

Comment:
AFAIK there's no k8s backend specific stuff in this function and we can simply keep using c.server.getNeighborConfigs. Am I missing something?

Reply:
bgpconfig is information obtained from k8s backend.